File: src/main/java/com/uiptv/db/SeriesChannelDb.java
package com.uiptv.db;

import com.uiptv.model.Account;
import com.uiptv.model.Channel;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import static com.uiptv.db.DatabaseUtils.DbTable.SERIES_CHANNEL_TABLE;
import static com.uiptv.db.DatabaseUtils.insertTableSql;
import static com.uiptv.db.SQLConnection.connect;

public class SeriesChannelDb extends BaseDb {
    private static SeriesChannelDb instance;

    public SeriesChannelDb() {
        super(SERIES_CHANNEL_TABLE);
    }

    public static synchronized SeriesChannelDb get() {
        if (instance == null) {
            instance = new SeriesChannelDb();
        }
        return instance;
    }

    public List<Channel> getChannels(Account account, String categoryId) {
        return getAll(" WHERE accountId=? AND categoryId=?", new String[]{account.getDbId(), categoryId});
    }

    public List<Channel> getChannelsBySeriesIds(Account account, List<String> seriesIds) {
        if (account == null || seriesIds == null || seriesIds.isEmpty()) {
            return List.of();
        }
        List<String> filtered = seriesIds.stream()
                .filter(id -> id != null && !id.isBlank())
                .distinct()
                .toList();
        if (filtered.isEmpty()) {
            return List.of();
        }
        String placeholders = String.join(",", java.util.Collections.nCopies(filtered.size(), "?"));
        String where = " WHERE accountId=? AND channelId IN (" + placeholders + ")";
        List<String> params = new ArrayList<>(filtered.size() + 1);
        params.add(account.getDbId());
        params.addAll(filtered);
        return getAll(where, params.toArray(new String[0]));
    }

    public boolean isFresh(Account account, String categoryId, long maxAgeMs) {
        String sql = "SELECT MAX(cachedAt) FROM " + SERIES_CHANNEL_TABLE.getTableName() + " WHERE accountId=? AND categoryId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, account.getDbId());
            statement.setString(2, categoryId);
            try (ResultSet rs = statement.executeQuery()) {
                if (rs.next()) {
                    long cachedAt = rs.getLong(1);
                    return cachedAt > 0 && (System.currentTimeMillis() - cachedAt) <= maxAgeMs;
                }
            }
        } catch (SQLException ignored) {
        }
        return false;
    }

    public void saveAll(List<Channel> channels, String categoryId, Account account) {
        deleteByAccountAndCategory(account.getDbId(), categoryId);
        long cachedAt = System.currentTimeMillis();
        channels.forEach(c -> insert(c, categoryId, account, cachedAt));
    }

    public void deleteByAccount(String accountId) {
        String sql = "DELETE FROM " + SERIES_CHANNEL_TABLE.getTableName() + " WHERE accountId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute delete query", e);
        }
    }

    private void deleteByAccountAndCategory(String accountId, String categoryId) {
        String sql = "DELETE FROM " + SERIES_CHANNEL_TABLE.getTableName() + " WHERE accountId=? AND categoryId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.setString(2, categoryId);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute delete query", e);
        }
    }

    private void insert(Channel channel, String categoryId, Account account, long cachedAt) {
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(insertTableSql(SERIES_CHANNEL_TABLE))) {
            statement.setString(1, channel.getChannelId());
            statement.setString(2, categoryId);
            statement.setString(3, account.getDbId());
            statement.setString(4, channel.getName());
            statement.setString(5, channel.getNumber());
            statement.setString(6, channel.getCmd());
            statement.setString(7, channel.getCmd_1());
            statement.setString(8, channel.getCmd_2());
            statement.setString(9, channel.getCmd_3());
            statement.setString(10, channel.getLogo());
            statement.setInt(11, channel.getCensored());
            statement.setInt(12, channel.getStatus());
            statement.setInt(13, channel.getHd());
            statement.setString(14, channel.getDrmType());
            statement.setString(15, channel.getDrmLicenseUrl());
            statement.setString(16, channel.getClearKeysJson());
            statement.setString(17, channel.getInputstreamaddon());
            statement.setString(18, channel.getManifestType());
            statement.setString(19, channel.getExtraJson());
            statement.setLong(20, cachedAt);
            statement.execute();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to execute insert query", e);
        }
    }

    @Override
    Channel populate(ResultSet resultSet) {
        Channel c = new Channel(
                nullSafeString(resultSet, "channelId"),
                nullSafeString(resultSet, "name"),
                nullSafeString(resultSet, "number"),
                nullSafeString(resultSet, "cmd"),
                nullSafeString(resultSet, "cmd_1"),
                nullSafeString(resultSet, "cmd_2"),
                nullSafeString(resultSet, "cmd_3"),
                nullSafeString(resultSet, "logo"),
                safeInteger(resultSet, "censored"),
                safeInteger(resultSet, "status"),
                safeInteger(resultSet, "hd"),
                nullSafeString(resultSet, "drmType"),
                nullSafeString(resultSet, "drmLicenseUrl"),
                null,
                nullSafeString(resultSet, "inputstreamaddon"),
                nullSafeString(resultSet, "manifestType")
        );
        c.setDbId(nullSafeString(resultSet, "id"));
        c.setCategoryId(nullSafeString(resultSet, "categoryId"));
        c.setClearKeysJson(nullSafeString(resultSet, "clearKeysJson"));
        c.setExtraJson(nullSafeString(resultSet, "extraJson"));
        return c;
    }
}

File: server/src/main/kotlin/com/uiptv/db/SeriesChannelDb.kt
package com.uiptv.db

import com.uiptv.model.Account
import com.uiptv.model.Channel
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SqlExpressionBuilder.inList
import org.jetbrains.exposed.sql.Table
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.batchInsert
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.statements.UpdateBuilder

class SeriesChannelDb private constructor() : ExposedCrudRepository<String, Channel>() {
    companion object {
        private val instance = SeriesChannelDb()

        @JvmStatic
        fun get(): SeriesChannelDb = instance
    }

    override fun findAll(): List<Channel> = query {
        SeriesChannelTable.selectAll().map(ResultRow::toSeriesChannel)
    }

    override fun findById(id: String): Channel? = query {
        id.toIntOrNull()
            ?.let { dbId -> SeriesChannelTable.selectAll().where { SeriesChannelTable.id eq dbId }.limit(1).firstOrNull() }
            ?.toSeriesChannel()
    }

    override fun save(entity: Channel): Channel {
        val dbId = entity.dbId?.toIntOrNull()
        query {
            if (dbId == null) {
                val insertedId = SeriesChannelTable.insert { row -> row.write(entity, 0L, "", "") }[SeriesChannelTable.id]
                entity.dbId = insertedId.toString()
            } else {
                SeriesChannelTable.deleteWhere { SeriesChannelTable.id eq dbId }
                val insertedId = SeriesChannelTable.insert { row -> row.write(entity, 0L, entity.categoryId.orEmpty(), "") }[SeriesChannelTable.id]
                entity.dbId = insertedId.toString()
            }
        }
        return entity
    }

    override fun deleteById(id: String) {
        val dbId = id.toIntOrNull() ?: return
        query {
            SeriesChannelTable.deleteWhere { SeriesChannelTable.id eq dbId }
        }
    }

    fun getChannels(account: Account, categoryId: String): List<Channel> = query {
        SeriesChannelTable.selectAll()
            .where { (SeriesChannelTable.accountId eq account.dbId.orEmpty()) and (SeriesChannelTable.categoryId eq categoryId) }
            .map(ResultRow::toSeriesChannel)
    }

    fun getChannelsBySeriesIds(account: Account?, seriesIds: List<String?>?): List<Channel> {
        if (account == null || seriesIds.isNullOrEmpty()) {
            return emptyList()
        }
        val filtered = seriesIds.filterNotNull().filter(String::isNotBlank).distinct()
        if (filtered.isEmpty()) {
            return emptyList()
        }
        return query {
            SeriesChannelTable.selectAll()
                .where {
                    (SeriesChannelTable.accountId eq account.dbId.orEmpty()) and
                        (SeriesChannelTable.channelId inList filtered)
                }
                .map(ResultRow::toSeriesChannel)
        }
    }

    fun isFresh(account: Account, categoryId: String, maxAgeMs: Long): Boolean = query {
        if (maxAgeMs <= 0) {
            false
        } else {
            val latest = SeriesChannelTable.selectAll()
                .where { (SeriesChannelTable.accountId eq account.dbId.orEmpty()) and (SeriesChannelTable.categoryId eq categoryId) }
                .map { it[SeriesChannelTable.cachedAt] }
                .maxOrNull() ?: 0L
            latest > 0 && (System.currentTimeMillis() - latest) <= maxAgeMs
        }
    }

    fun saveAll(channels: List<Channel>, categoryId: String, account: Account) {
        val accountId = account.dbId.orEmpty()
        query {
            SeriesChannelTable.deleteWhere {
                (SeriesChannelTable.accountId eq accountId) and (SeriesChannelTable.categoryId eq categoryId)
            }
            if (channels.isNotEmpty()) {
                val cachedAt = System.currentTimeMillis()
                SeriesChannelTable.batchInsert(channels) { channel ->
                    write(channel, cachedAt, categoryId, accountId)
                }
            }
        }
    }

    fun deleteByAccount(accountId: String) {
        query {
            SeriesChannelTable.deleteWhere { SeriesChannelTable.accountId eq accountId }
        }
    }
}

private object SeriesChannelTable : Table(DatabaseUtils.DbTable.SERIES_CHANNEL_TABLE.tableName) {
    val id = integer("id").autoIncrement()
    val channelId = text("channelId")
    val categoryId = text("categoryId").nullable()
    val accountId = text("accountId").nullable()
    val name = text("name").nullable()
    val number = text("number").nullable()
    val cmd = text("cmd").nullable()
    val cmd1 = text("cmd_1").nullable()
    val cmd2 = text("cmd_2").nullable()
    val cmd3 = text("cmd_3").nullable()
    val logo = text("logo").nullable()
    val censored = integer("censored")
    val status = integer("status")
    val hd = integer("hd")
    val drmType = text("drmType").nullable()
    val drmLicenseUrl = text("drmLicenseUrl").nullable()
    val clearKeysJson = text("clearKeysJson").nullable()
    val inputstreamaddon = text("inputstreamaddon").nullable()
    val manifestType = text("manifestType").nullable()
    val extraJson = text("extraJson").nullable()
    val cachedAt = long("cachedAt")

    override val primaryKey = PrimaryKey(id)
}

private fun ResultRow.toSeriesChannel(): Channel = Channel(
    dbId = this[SeriesChannelTable.id].toString(),
    channelId = this[SeriesChannelTable.channelId],
    categoryId = this[SeriesChannelTable.categoryId],
    name = this[SeriesChannelTable.name],
    number = this[SeriesChannelTable.number],
    cmd = this[SeriesChannelTable.cmd],
    cmd_1 = this[SeriesChannelTable.cmd1],
    cmd_2 = this[SeriesChannelTable.cmd2],
    cmd_3 = this[SeriesChannelTable.cmd3],
    logo = this[SeriesChannelTable.logo],
    extraJson = this[SeriesChannelTable.extraJson],
    censored = this[SeriesChannelTable.censored],
    status = this[SeriesChannelTable.status],
    hd = this[SeriesChannelTable.hd],
    drmType = this[SeriesChannelTable.drmType],
    drmLicenseUrl = this[SeriesChannelTable.drmLicenseUrl],
    clearKeysJson = this[SeriesChannelTable.clearKeysJson],
    inputstreamaddon = this[SeriesChannelTable.inputstreamaddon],
    manifestType = this[SeriesChannelTable.manifestType]
)

private fun UpdateBuilder<*>.write(channel: Channel, cachedAt: Long, categoryId: String, accountId: String) {
    this[SeriesChannelTable.channelId] = channel.channelId.orEmpty()
    this[SeriesChannelTable.categoryId] = categoryId
    this[SeriesChannelTable.accountId] = accountId
    this[SeriesChannelTable.name] = channel.name
    this[SeriesChannelTable.number] = channel.number
    this[SeriesChannelTable.cmd] = channel.cmd
    this[SeriesChannelTable.cmd1] = channel.cmd_1
    this[SeriesChannelTable.cmd2] = channel.cmd_2
    this[SeriesChannelTable.cmd3] = channel.cmd_3
    this[SeriesChannelTable.logo] = channel.logo
    this[SeriesChannelTable.censored] = channel.censored
    this[SeriesChannelTable.status] = channel.status
    this[SeriesChannelTable.hd] = channel.hd
    this[SeriesChannelTable.drmType] = channel.drmType
    this[SeriesChannelTable.drmLicenseUrl] = channel.drmLicenseUrl
    this[SeriesChannelTable.clearKeysJson] = channel.clearKeysJson
    this[SeriesChannelTable.inputstreamaddon] = channel.inputstreamaddon
    this[SeriesChannelTable.manifestType] = channel.manifestType
    this[SeriesChannelTable.extraJson] = channel.extraJson
    this[SeriesChannelTable.cachedAt] = cachedAt
}
